### PR TITLE
feat: Use Appender interface

### PR DESCRIPTION
**Pull request by disq**
Avoid writing to parquet if we can (appends only)

~Using upstream driver has some shortcomings from the get go: The interface decides on the schema based on passed types, so having untyped `nil`s don't help. But typed nils are also not handled. The [fork/PR](https://github.com/marcboeker/go-duckdb/pull/150) fixes these issues but I still get 0 rows on the test db (or motherduck)~ The PR is now merged upstream and tagged as 1.6.1.

~I had to use my [fork](https://github.com/disq/go-duckdb/tree/feat/allow-nulls-in-appender-first-row) to get around nils, another way to this would be if the `appender.initColTypes` was exported, then we could just pass on concrete types for init and then use nullable data as normal.~

~Another issue: We create JSON types as `json` but appender doesn't know about it. Sending `string` instead (which is what the json alias points to internally in duckdb, varchar) doesn't work. Might need the newer appender implementation for this to go forward.~

Example config:
```yaml
kind: source
spec:
  name: test
  path: cloudquery/test
  registry: cloudquery
  version: v4.0.8
#  tables: ["*"]
  tables: [ "test_some_table" ]
  destinations: ["duckdb"]
  spec:
    num_clients: 1
    num_rows: 4
    num_sub_rows: 1
#    num_clients: 10
#    num_rows: 10000
    num_sub_rows: 0
#    num_sub_rows: 10
---
kind: destination
spec:
  name: duckdb
  version: v0.0.0
  registry: grpc
  path: localhost:7777
#  write_mode: "overwrite"
  write_mode: "append"
  spec:
    connection_string: tmp_duck.db
#    connection_string: md:my_db
    # Optional parameters
    batch_size: 10000
#    batch_size_bytes: 14194304 # 14 MiB
    # debug: false
```

### CODE REVIEW

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+		case arrow.STRING:
+			return ptrOf[string]()
+		case arrow.BINARY, arrow.LARGE_BINARY, arrow.FIXED_SIZE_BINARY:
+			return ptrOf[[]byte]()
+		default:
+			return ptrOf[string]()
+		}
+	}
+
+	if arrow.TypeEqual(arr.DataType(), types.NewUUIDType()) {
```
Comment (disq) [inline plugins/destination/duckdb/client/transform_gotype.go:53]: This is probably ineffective

File: plugins/destination/duckdb/client/client.go [hunk 2]
```diff
@@ -56,12 +57,11 @@ func New(ctx context.Context, logger zerolog.Logger, spec []byte, _ plugin.NewCl
 			internalPlugin.Team, internalPlugin.Kind, internalPlugin.Name, strings.TrimPrefix(internalPlugin.Version, "v"), runtime.GOOS, runtime.GOARCH)
 	}
 
-	c.connector, err = duckdb.NewConnector(c.spec.ConnectionString, nil)
+	dbconnector, err := duckdb.NewConnector(c.spec.ConnectionString, nil)
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/client.go:60]: did you forget to init `c.conn` somewhere around here?

File: plugins/destination/duckdb/client/client.go [hunk 3]
```diff
@@ -88,8 +86,20 @@ func (c *Client) Close(ctx context.Context) error {
...
+	err := c.db.Close()
 	c.db = nil
+
+	if c.conn != nil {
+		err2 := c.conn.Close()
+		if err2 != nil {
+			fmt.Println("conn close failed: ", err2)
+		}
+		c.conn = nil
+		if err == nil {
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/client.go:98]: I'd just go with simple
```go
return errors.Join(c.db.Close(), c.conn.Close())
```

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+)
+
+func getValue(arr arrow.Array, i int) any {
+	if !arr.IsValid(i) {
+		if arrow.TypeEqual(arr.DataType(), types.NewUUIDType()) {
+			return duckdb.UUID{}
+		}
+
+		// return typed nil
+		switch arr.DataType().ID() {
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/transform_gotype.go:20]: IMO it's better to do a type switch instead (see ch dst):
```go
switch dt := arr.DataType().(type)
```
with stuff like `*arrow.TImestampType`

File: plugins/destination/duckdb/client/transform_gotype.go (hunk 1 above)
Comment (candiduslynx) [reply plugins/destination/duckdb/client/transform_gotype.go:54]: can be incorporated with the type switch as `case *types.UUIDType:`

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/marcboeker/go-duckdb"
+)
+
+func getValue(arr arrow.Array, i int) any {
+	if !arr.IsValid(i) {
+		if arrow.TypeEqual(arr.DataType(), types.NewUUIDType()) {
+			return duckdb.UUID{}
+		}
+
+		// return typed nil
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/transform_gotype.go:19]: should be moved to a separate func

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+			return ptrOf[string]()
+		}
+	}
+
+	if arrow.TypeEqual(arr.DataType(), types.NewUUIDType()) {
+		v, _ := arr.(*types.UUIDArray).Value(i).MarshalBinary()
+		return duckdb.UUID(v)
+	}
+
+	switch arr.DataType().ID() {
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/transform_gotype.go:59]: use type switch here, too

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	var res [][]driver.Value
+	for i := int64(0); i < record.NumRows(); i++ {
+		nc := record.NumCols()
+		row := make([]driver.Value, nc)
+		for j := 0; int64(j) < nc; j++ {
+			row[j] = getValue(record.Column(j), int(i))
+		}
+		res = append(res, row)
+	}
+	return res
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/transform_gotype.go:109]: prealloc should work fine here
```suggestion
	res := make([][]driver.Value, record.NumRows())
	nc := record.NumCols()
	for i := range res {
		res[i] = make([]driver.Value, nc)
	}

	for j := int64(0); j < nc; j++ {
		col := record.Column(j)
		for i := range res {
			 res[i][j] = getValue(col, i)
		}
	}
	return res
```

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+func ptrOf[T any]() *T {
+	var val *T
+	return val
+}
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/transform_gotype.go:115]: I wonder if thsi can be simplified via
```suggestion
func nilPtrOf[T any]() *T {
	return nil
}
```

File: plugins/destination/duckdb/client/write.go [hunk 9]
```diff
@@ -125,12 +127,60 @@ func (c *Client) Write(ctx context.Context, msgs <-chan message.WriteMessage) er
...
+	fmt.Println("appendRows", table.Name)
+	appender, err := duckdb.NewAppenderFromConn(c.conn, "", table.Name)
+	if err != nil {
+		return fmt.Errorf("failed to create appender for %s: %w", table.Name, err)
+	}
+
+	for _, msg := range msgs {
+		arr := transformRecordToGoType(msg.Record)
+		for i := range arr {
+			b, _ := json.Marshal(arr[i])
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/write.go:140]: err shouldn't be ignored IMO

**Review by candiduslynx — CHANGES_REQUESTED** [review]
some comments

File: plugins/destination/duckdb/client/client.go (hunk 2 above)
Comment (disq) [reply plugins/destination/duckdb/client/client.go:60]: No, c.conn is for appender.

File: plugins/destination/duckdb/client/write.go (hunk 9 above)
Comment (disq) [reply plugins/destination/duckdb/client/write.go:140]: it's just for `fmt.Println("appendRows", string(b))`

File: plugins/destination/duckdb/client/transform_gotype.go (hunk 8 above)
Comment (disq) [reply plugins/destination/duckdb/client/transform_gotype.go:118]: it works <3

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 10]
```diff
@@ -42,6 +43,20 @@ func getTypedNilValue(arr arrow.Array) any {
 		return nilPtrOf[string]()
 	case *arrow.BinaryType, *arrow.LargeBinaryType, *arrow.FixedSizeBinaryType:
 		return nilPtrOf[[]byte]()
+	case *arrow.ListType:
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/transform_gotype.go:46]: can we use [`arrow.ListLikeType`](https://pkg.go.dev/github.com/apache/arrow/go/v15@v15.0.0/arrow#ListLikeType) interface to match both lists (NB: also matches maps)?

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 11]
```diff
@@ -63,6 +78,12 @@ func getValue(arr arrow.Array, i int, firstRow bool) any {
 		ts := arr.(*array.Timestamp)
 		timeUnit := ts.DataType().(*arrow.TimestampType).Unit
 		return ts.Value(i).ToTime(timeUnit)
+	case *arrow.Date32Type:
+		ts := arr.(*array.Date32)
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/transform_gotype.go:82]: don't switch over `arr.DataType().(type)`, use `switch arr := arr.(type)`

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 12]
```diff
@@ -42,6 +43,20 @@ func getTypedNilValue(arr arrow.Array) any {
...
+	case *arrow.ListType:
+		arr := arr.(*array.List)
+		v := getTypedNilValue(arr.ListValues())
+		return reflect.New(reflect.SliceOf(reflect.TypeOf(v))).Interface()
+	case *arrow.LargeListType:
+		arr := arr.(*array.LargeList)
+		v := getTypedNilValue(arr.ListValues())
+		return reflect.New(reflect.SliceOf(reflect.TypeOf(v))).Interface()
+	case *arrow.StructType:
+		// Can't create a Go struct dynamically and maps are unsupported: use string
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/transform_gotype.go:55]: can we do JSON?

Comment (disq) [reply plugins/destination/duckdb/client/transform_gotype.go:55]: As long as it matches what read() (which uses parquet) expects in type and data, we can try.

File: plugins/destination/duckdb/client/transform_gotype.go (hunk 10 above)
Comment (disq) [reply plugins/destination/duckdb/client/transform_gotype.go:46]: yeah it matches maps as well, but we can order it carefully

Comment (disq) [reply plugins/destination/duckdb/client/transform_gotype.go:46]: Done in https://github.com/cloudquery/cloudquery/pull/16668/commits/3ce23706feccf97e67ba6820483d56044115844d

File: plugins/destination/duckdb/client/transform_gotype.go (hunk 12 above)
Comment (disq) [reply plugins/destination/duckdb/client/transform_gotype.go:55]: Turns out ValueStr does exactly that for Struct:

```go
func (a *Struct) ValueStr(i int) string {
	if a.IsNull(i) {
		return NullValueStr
	}

	data, err := json.Marshal(a.GetOneForMarshal(i))
	if err != nil {
		panic(err)
	}
	return string(data)
}
```

File: plugins/destination/duckdb/client/transform_gotype.go (hunk 10 above)
Comment (candiduslynx) [reply plugins/destination/duckdb/client/transform_gotype.go:46]: I still see it here, maybe it was done only in the values but not in nils?

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+		return nilPtrOf[string]()
+	case *arrow.MapType:
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/transform_gotype.go:57]: should be moved before `arrow.ListLikeType` case

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+				val.Index(i).SetZero() // not sure if this is necessary
+				continue
+			}
+			// slice of pointers, make everything a pointer
+			sv := reflect.ValueOf(getValue(slice, i))
+			psv := reflect.New(sv.Type())
+			psv.Elem().Set(sv)
+			val.Index(i).Set(psv)
+		}
+		return val.Interface()
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/transform_gotype.go:132]: I wonder if returning `[]any` with values would work better here...

File: plugins/destination/duckdb/client/transform_gotype.go (hunk 13 above)
Comment (disq) [reply plugins/destination/duckdb/client/transform_gotype.go]: Fixed in https://github.com/cloudquery/cloudquery/pull/16668/commits/34abf5cb6234ebd7cbf36ac97b05fe9f2b564723

File: plugins/destination/duckdb/client/transform_gotype.go (hunk 14 above)
Comment (disq) [reply plugins/destination/duckdb/client/transform_gotype.go:126]: smart, but no cigar:
> 10:04AM ERR failed to write batch error="failed to append row to test_testdata_table: type mismatch for column 42: \nexpected: []int64, actual: []interface {}" duration=1.473458 len=1 module=duckdb-dest table=test_testdata_table

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 15]
```diff
@@ -12,51 +12,45 @@ import (
 )
 
 func getTypedNilValue(arr arrow.Array) any {
-	switch arr.DataType().(type) {
-	case *types.UUIDType:
+	switch arr.(type) {
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/transform_gotype.go:15]:
```suggestion
	switch arr := arr.(type) {
```

Comment (disq) [reply plugins/destination/duckdb/client/transform_gotype.go:15]: if I do that then I can't merge cases.

Comment (candiduslynx) [reply plugins/destination/duckdb/client/transform_gotype.go:15]: why? you definitely can

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	for i, col := range arrowCols { // i: arrow column index, col: arrow column name
+		j, ok := columnMap[col] // look up the column index in the destination table
+		if !ok {
+			continue
+		}
+		arr := record.Column(i)
+		for k := range res {
+			res[k][j] = getValue(arr, k)
```
Comment (disq) [inline plugins/destination/duckdb/client/transform_gotype.go:148]: @candiduslynx juicy

File: plugins/destination/duckdb/go.mod [hunk 17]
```diff
@@ -134,3 +134,11 @@ require (
 
 // github.com/cloudquery/jsonschema @ cqmain
 replace github.com/invopop/jsonschema => github.com/cloudquery/jsonschema v0.0.0-20240202134451-d771afde32fb
+
+// replace github.com/marcboeker/go-duckdb => ../../../../../disq/go-duckdb
```
Comment (disq) [inline plugins/destination/duckdb/go.mod:138]: These will need removing once we finalize this PR

File: plugins/destination/duckdb/client/client_test.go [hunk 18]
```diff
@@ -40,6 +49,9 @@ func TestPlugin(t *testing.T) {
 			SkipIntervals: true,
 			// not supported in duckDB for now
 			SkipLargeTypes: true,
+			// not supported in Appender:
+			SkipMaps:    true,
+			SkipStructs: true,
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/client_test.go:54]: I wonder if we can remove this & use string/JSON?

File: plugins/destination/duckdb/client/write.go [hunk 19]
```diff
@@ -139,14 +186,16 @@ func (c *Client) WriteTableBatch(ctx context.Context, name string, msgs message.
...
-	if len(table.PrimaryKeys()) == 0 {
-		copyStart := time.Now()
-		defer func() {
-			c.logger.Debug().Str("table", table.Name).Str("duration", time.Since(copyStart).String()).Msg("copy file to table")
-		}()
-		return c.copyFromFile(ctx, name, tmpFile, table)
-	}
-
+	/*
+		if len(table.PrimaryKeys()) == 0 {
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/write.go:190]: remove?

File: plugins/destination/duckdb/client/types.go [hunk 20]
```diff
@@ -38,16 +38,21 @@ func transformTypeForWriting(dt arrow.DataType) arrow.DataType {
...
 	case *arrow.StructType:
-		builder := new(strings.Builder)
-		builder.WriteString("struct(")
-		for i, field := range dt.Fields() {
-			if i > 0 {
-				builder.WriteString(", ")
+		// use json for struct types for appender compatibility
+		return "json"
+	/*
+		case *arrow.StructType:
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/types.go:44]: remove obsolete code?

File: plugins/destination/duckdb/client/transform_gotype.go [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+		for i := 0; i < slice.Len(); i++ {
+			if slice.IsNull(i) {
+				val.Index(i).SetZero() // not sure if this is necessary
+				continue
+			}
+			// slice of pointers, make everything a pointer
+			sv := reflect.ValueOf(getValue(slice, i))
+			psv := reflect.New(sv.Type())
+			psv.Elem().Set(sv)
+			val.Index(i).Set(psv)
```
Comment (candiduslynx) [inline plugins/destination/duckdb/client/transform_gotype.go:124]: please extract this into a separate func

**Review by candiduslynx — APPROVED** [review]

**Comment by disq** [comment]
I'm going to pick this up after https://github.com/cloudquery/cloudquery/pull/16772 is merged

File: plugins/destination/duckdb/client/client_test.go (hunk 18 above)
Comment (disq) [reply plugins/destination/duckdb/client/client_test.go]: Struct as JSON done in https://github.com/cloudquery/cloudquery/pull/16668/commits/ecaf3a261cd96eaf8cf0e62cc7b80003674c5553 will now work on maps

File: plugins/destination/duckdb/client/write.go (hunk 19 above)
Comment (disq) [reply plugins/destination/duckdb/client/write.go:189]: I uncomment it from time to time to test direct parquet copys.

File: plugins/destination/duckdb/client/client_test.go (hunk 18 above)
Comment (disq) [reply plugins/destination/duckdb/client/client_test.go]: Maps as JSON done in https://github.com/cloudquery/cloudquery/pull/16668/commits/6eaf9da7f2ad42923ed578d05531230276168661 PTAL

**Pull request merged**